File: src/exdl_alc.erl
%%  Copyright (c) 2008 David Koch
%%
%%  See the file "license.terms" for information on usage and redistribution
%%  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
%% 
%%     $Id: exdl_alc.erl,v 1.2 2008/11/06 07:49:22 kochise Exp $
%%

%%%----------------------------------------------------------------------
%%% File     : exdl_alc.erl
%%% Author   : David Koch <david.koch@libertysurf.fr>
%%% Purpose  : Implements alc (OpenAL Complementary lib)
%%% Created  : 6 Nov 2008 by David Koch <david.koch@libertysurf.fr>
%%% Modified : 
%%%----------------------------------------------------------------------

-module(exdl_alc).

-include("exdl.hrl").

-ifdef(COMPILE_AL).

%% 1- EXPORT

%-compile(export_all).

%% Context Management

-export( [ createContext/2
         , makeContextCurrent/1
         , processContext/1
         , suspendContext/1
         , destroyContext/1
         , getCurrentContext/0
         , getContextsDevice/1
         ]
       ).

%% Device Management

-export( [ openDevice/1
         , closeDevice/1
         ]
       ).

%% Error support

-export( [ getError/1
         ]
       ).

%% Extension support

-export( [ isExtensionPresent/2
         , getProcAddress/2
         , getEnumValue/2
         ]
       ).

%% Query functions

-export( [ getString/2
         , getIntegerv/3
         ]
       ).

%% Capture functions

-export( [ captureOpenDevice/4
         , captureCloseDevice/1
         , captureStart/1
         , captureStop/1
         , captureSamples/3
         ]
       ).

%% 2- INCLUDE

-include("exdl_util.hrl").

-include("exdl_alc.hrl").

%% 3- IMPORT

-import( exdl_driver
       , [ call/2
         , cast/2
         ]
       ).

-import( exdl_util
       , [ bin2list/2
         , bin2list/3
         , term2bin/2
         , term2bin/3
         , matrix2bin/2
         ]
       ).

%% 4- FUNCTIONS

%% Context Management

%% Func:       createContext
%% Args:       Device, Attrlist
%% Returns:    Exdl
%% C-API func: ALCcontext* alcCreateContext(ALCdevice* device, const ALCint* attrlist)

createContext(Device, Attrlist)
 -> case
      call
      ( ?alcCreateContext
      , <<Device:?_PTR
        , Attrlist:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       makeContextCurrent
%% Args:       Context
%% Returns:    Exdl
%% C-API func: ALCboolean alcMakeContextCurrent(ALCcontext* context)

makeContextCurrent(Context)
 -> case
      call
      ( ?alcMakeContextCurrent
      , <<Context:?_PTR
        >>
      )
    of
      <<Exdl:?ALCboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       processContext
%% Args:       Context
%% Returns:    ok
%% C-API func: void alcProcessContext(ALCcontext* context)

processContext(Context)
 -> cast
    ( ?alcProcessContext
    , <<Context:?_PTR
      >>
    )
.

%% Func:       suspendContext
%% Args:       Context
%% Returns:    ok
%% C-API func: void alcSuspendContext(ALCcontext* context)

suspendContext(Context)
 -> cast
    ( ?alcSuspendContext
    , <<Context:?_PTR
      >>
    )
.

%% Func:       destroyContext
%% Args:       Context
%% Returns:    ok
%% C-API func: void alcDestroyContext(ALCcontext* context)

destroyContext(Context)
 -> cast
    ( ?alcDestroyContext
    , <<Context:?_PTR
      >>
    )
.

%% Func:       getCurrentContext
%% Args:       none
%% Returns:    Exdl
%% C-API func: ALCcontext* alcGetCurrentContext(void)

getCurrentContext()
 -> case
      call
      ( ?alcGetCurrentContext
      , [
        ]
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getContextsDevice
%% Args:       Context
%% Returns:    Exdl
%% C-API func: ALCdevice* alcGetContextsDevice(ALCcontext* context)

getContextsDevice(Context)
 -> case
      call
      ( ?alcGetContextsDevice
      , <<Context:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Device Management

%% Func:       openDevice
%% Args:       Devicename
%% Returns:    Exdl
%% C-API func: ALCdevice* alcOpenDevice(const ALCchar* devicename)

openDevice(Devicename)
 -> case
      call
      ( ?alcOpenDevice
      , <<Devicename:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       closeDevice
%% Args:       Device
%% Returns:    Exdl
%% C-API func: ALCboolean alcCloseDevice(ALCdevice* device)

closeDevice(Device)
 -> case
      call
      ( ?alcCloseDevice
      , <<Device:?_PTR
        >>
      )
    of
      <<Exdl:?ALCboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Error support

%% Func:       getError
%% Args:       Device
%% Returns:    Exdl
%% C-API func: ALCenum alcGetError(ALCdevice* device)

getError(Device)
 -> case
      call
      ( ?alcGetError
      , <<Device:?_PTR
        >>
      )
    of
      <<Exdl:?ALCenum
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Extension support

%% Func:       isExtensionPresent
%% Args:       Device, Extname
%% Returns:    Exdl
%% C-API func: ALCboolean alcIsExtensionPresent(ALCdevice* device, const ALCchar* extname)

isExtensionPresent(Device, Extname)
 -> case
      call
      ( ?alcIsExtensionPresent
      , <<Device:?_PTR
        , Extname:?_PTR
        >>
      )
    of
      <<Exdl:?ALCboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getProcAddress
%% Args:       Device, Funcname
%% Returns:    Exdl
%% C-API func: void* alcGetProcAddress(ALCdevice* device, const ALCchar* funcname)

getProcAddress(Device, Funcname)
 -> case
      call
      ( ?alcGetProcAddress
      , <<Device:?_PTR
        , Funcname:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getEnumValue
%% Args:       Device, Enumname
%% Returns:    Exdl
%% C-API func: ALCenum alcGetEnumValue(ALCdevice* device, const ALCchar* enumname)

getEnumValue(Device, Enumname)
 -> case
      call
      ( ?alcGetEnumValue
      , <<Device:?_PTR
        , Enumname:?_PTR
        >>
      )
    of
      <<Exdl:?ALCenum
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Query functions

%% Func:       getString
%% Args:       Device, Param
%% Returns:    Exdl
%% C-API func: const ALCchar* alcGetString(ALCdevice* device, ALCenum param)

getString(Device, Param)
 -> case
      call
      ( ?alcGetString
      , <<Device:?_PTR
        , Param:?ALCenum
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?String
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getIntegerv
%% Args:       Device, Param, Size
%% Returns:    [ Data ]
%% C-API func: void alcGetIntegerv(ALCdevice* device, ALCenum param, ALCsizei size, ALCint* data)

getIntegerv(Device, Param, Size)
 -> case
      call
      ( ?alcGetIntegerv
      , <<Device:?_PTR
        , Param:?ALCenum
        , Size:?ALCsizei
        >>
      )
    of
      <<Data:?ALCint
      >>
        -> [ Data ]
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Capture functions

%% Func:       captureOpenDevice
%% Args:       Devicename, Frequency, Format, Buffersize
%% Returns:    Exdl
%% C-API func: ALCdevice* alcCaptureOpenDevice(const ALCchar* devicename, ALCuint frequency, ALCenum format, ALCsizei buffersize)

captureOpenDevice(Devicename, Frequency, Format, Buffersize)
 -> case
      call
      ( ?alcCaptureOpenDevice
      , <<Devicename:?_PTR
        , Frequency:?ALCuint
        , Format:?ALCenum
        , Buffersize:?ALCsizei
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       captureCloseDevice
%% Args:       Device
%% Returns:    Exdl
%% C-API func: ALCboolean alcCaptureCloseDevice(ALCdevice* device)

captureCloseDevice(Device)
 -> case
      call
      ( ?alcCaptureCloseDevice
      , <<Device:?_PTR
        >>
      )
    of
      <<Exdl:?ALCboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       captureStart
%% Args:       Device
%% Returns:    ok
%% C-API func: void alcCaptureStart(ALCdevice* device)

captureStart(Device)
 -> cast
    ( ?alcCaptureStart
    , <<Device:?_PTR
      >>
    )
.

%% Func:       captureStop
%% Args:       Device
%% Returns:    ok
%% C-API func: void alcCaptureStop(ALCdevice* device)

captureStop(Device)
 -> cast
    ( ?alcCaptureStop
    , <<Device:?_PTR
      >>
    )
.

%% Func:       captureSamples
%% Args:       Device, Buffer, Samples
%% Returns:    ok
%% C-API func: void alcCaptureSamples(ALCdevice* device, ALCvoid* buffer, ALCsizei samples)

captureSamples(Device, Buffer, Samples)
 -> cast
    ( ?alcCaptureSamples
    , <<Device:?_PTR
      , Buffer:?_PTR
      , Samples:?ALCsizei
      >>
    )
.

-endif. % COMPILE_AL

File: priv/def/test/exdl_alc.c
/*
 *  Copyright (c) 2012 David Koch
 *
 *  See the file "license.terms" for information on usage and redistribution
 *  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 *     $Id: exdl_alc.c,v 1.5 2005/06/29 14:45:38 dgud Exp $
 */

#include "exdl.h"

#ifdef COMPILE_ALC

/* Context Management */

EXDL_API(ealc_createContext)
{
  /* ALCcontext* alcCreateContext(ALCdevice* device, const ALCint* attrlist) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*,    device)
    STACK_FMT_DATA(ALCint,        attrlist[])
  STACK_FMT_END()

  CATCH_RESULT(ALCcontext*)

    alcCreateContext
    ( STACK_GET_DATA(device)
    , STACK_GET_ADDR(attrlist)
    );

  THROW_POINTER()
}

EXDL_API(ealc_makeContextCurrent)
{
  /* ALCboolean alcMakeContextCurrent(ALCcontext* context) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCcontext, context[])
  STACK_FMT_END()

  CATCH_RESULT(ALCboolean)

    alcMakeContextCurrent
    ( STACK_GET_ADDR(context)
    );

  THROW_RESULT(ALCboolean)
}

EXDL_API(ealc_processContext)
{
  /* void alcProcessContext(ALCcontext* context) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCcontext, context[])
  STACK_FMT_END()

    alcProcessContext
    ( STACK_GET_ADDR(context)
    );
}

EXDL_API(ealc_suspendContext)
{
  /* void alcSuspendContext(ALCcontext* context) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCcontext, context[])
  STACK_FMT_END()

    alcSuspendContext
    ( STACK_GET_ADDR(context)
    );
}

EXDL_API(ealc_destroyContext)
{
  /* void alcDestroyContext(ALCcontext* context) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCcontext, context[])
  STACK_FMT_END()

    alcDestroyContext
    ( STACK_GET_ADDR(context)
    );
}

EXDL_API(ealc_getCurrentContext)
{
  /* ALCcontext* alcGetCurrentContext(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(ALCcontext*)

    alcGetCurrentContext
    (
    );

  THROW_POINTER()
}

EXDL_API(ealc_getContextsDevice)
{
  /* ALCdevice* alcGetContextsDevice(ALCcontext* context) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCcontext*, context)
  STACK_FMT_END()

  CATCH_RESULT(ALCdevice*)

    alcGetContextsDevice
    ( STACK_GET_DATA(context)
    );

  THROW_POINTER()
}

/* Device Management */

EXDL_API(ealc_openDevice)
{
  /* ALCdevice* alcOpenDevice(const ALCchar* devicename) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCchar, devicename[])
  STACK_FMT_END()

  CATCH_RESULT(ALCdevice*)

    alcOpenDevice
    ( STACK_GET_ADDR(devicename)
    );

  THROW_POINTER()
}

EXDL_API(ealc_closeDevice)
{
  /* ALCboolean alcCloseDevice(ALCdevice* device) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCdevice, device[])
  STACK_FMT_END()

  CATCH_RESULT(ALCboolean)

    alcCloseDevice
    ( STACK_GET_ADDR(device)
    );

  THROW_RESULT(ALCboolean)
}

/* Error support */

EXDL_API(ealc_getError)
{
  /* ALCenum alcGetError(ALCdevice* device) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*, device)
  STACK_FMT_END()

  CATCH_RESULT(ALCenum)

    alcGetError
    ( STACK_GET_DATA(device)
    );

  THROW_RESULT(ALCenum)
}

/* Extension support */

EXDL_API(ealc_isExtensionPresent)
{
  /* ALCboolean alcIsExtensionPresent(ALCdevice* device, const ALCchar* extname) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*,     device)
    STACK_FMT_DATA(ALCchar,        extname[])
  STACK_FMT_END()

  CATCH_RESULT(ALCboolean)

    alcIsExtensionPresent
    ( STACK_GET_DATA(device)
    , STACK_GET_ADDR(extname)
    );

  THROW_RESULT(ALCboolean)
}

EXDL_API(ealc_getProcAddress)
{
  /* void* alcGetProcAddress(ALCdevice* device, const ALCchar* funcname) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*,     device)
    STACK_FMT_LPTR(ALCchar*,       funcname)
  STACK_FMT_END()

  CATCH_RESULT(void*)

    alcGetProcAddress
    ( STACK_GET_DATA(device)
    , STACK_GET_DATA(funcname)
    );

  THROW_POINTER()
}

EXDL_API(ealc_getEnumValue)
{
  /* ALCenum alcGetEnumValue(ALCdevice* device, const ALCchar* enumname) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*,     device)
    STACK_FMT_LPTR(ALCchar*,       enumname)
  STACK_FMT_END()

  CATCH_RESULT(ALCenum)

    alcGetEnumValue
    ( STACK_GET_DATA(device)
    , STACK_GET_DATA(enumname)
    );

  THROW_RESULT(ALCenum)
}

/* Query functions */

EXDL_API(ealc_getString)
{
  /* const ALCchar* alcGetString(ALCdevice* device, ALCenum param) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*, device)
    STACK_FMT_DATA(ALCenum,    param)
  STACK_FMT_END()

  CATCH_RESULT(const ALCchar*)

    alcGetString
    ( STACK_GET_DATA(device)
    , STACK_GET_DATA(param)
    );

  THROW_STRING()
}

EXDL_API(ealc_getIntegerv)
{
  /* void alcGetIntegerv(ALCdevice* device, ALCenum param, ALCsizei size, ALCint* data) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*, device)
    STACK_FMT_DATA(ALCenum,    param)
    STACK_FMT_DATA(ALCsizei,   size)
  STACK_FMT_END()

  FRAME_FMT_BEGIN()
    FRAME_FMT_DATA(ALCint,    data[1])
  FRAME_FMT_SIZE(STACK_GET_DATA(size) * sizeof(ALCint))

    alcGetIntegerv
    ( STACK_GET_DATA(device)
    , STACK_GET_DATA(param)
    , STACK_GET_DATA(size)
    , FRAME_GET_ADDR(data)
    );

  THROW_FRAME()
}

/* Capture functions */

EXDL_API(ealc_captureOpenDevice)
{
  /* ALCdevice* alcCaptureOpenDevice(const ALCchar* devicename, ALCuint frequency, ALCenum format, ALCsizei buffersize) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCchar*, devicename)
    STACK_FMT_DATA(ALCuint,  frequency)
    STACK_FMT_DATA(ALCenum,  format)
    STACK_FMT_DATA(ALCsizei, buffersize)
  STACK_FMT_END()

  CATCH_RESULT(ALCdevice*)

    alcCaptureOpenDevice
    ( STACK_GET_DATA(devicename)
    , STACK_GET_DATA(frequency)
    , STACK_GET_DATA(format)
    , STACK_GET_DATA(buffersize)
    );

  THROW_POINTER()
}

EXDL_API(ealc_captureCloseDevice)
{
  /* ALCboolean alcCaptureCloseDevice(ALCdevice* device) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCdevice, device[])
  STACK_FMT_END()

  CATCH_RESULT(ALCboolean)

    alcCaptureCloseDevice
    ( STACK_GET_ADDR(device)
    );

  THROW_RESULT(ALCboolean)
}

EXDL_API(ealc_captureStart)
{
  /* void alcCaptureStart(ALCdevice* device) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCdevice, device[])
  STACK_FMT_END()

    alcCaptureStart
    ( STACK_GET_ADDR(device)
    );
}

EXDL_API(ealc_captureStop)
{
  /* void alcCaptureStop(ALCdevice* device) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALCdevice, device[])
  STACK_FMT_END()

    alcCaptureStop
    ( STACK_GET_ADDR(device)
    );
}

EXDL_API(ealc_captureSamples)
{
  /* void alcCaptureSamples(ALCdevice* device, ALCvoid* buffer, ALCsizei samples) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(ALCdevice*, device)
    STACK_FMT_LPTR(ALCvoid*,   buffer)
    STACK_FMT_DATA(ALCsizei,   samples)
  STACK_FMT_END()

    alcCaptureSamples
    ( STACK_GET_DATA(device)
    , STACK_GET_DATA(buffer)
    , STACK_GET_DATA(samples)
    );
}

#endif // COMPILE_ALC
